_posts/2017-11-30-totw-130.md

```diff
@@ -0,0 +1,185 @@
+---
+title: "Tip of the Week #130: Namespace Naming"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/130
+type: markdown
+order: "130"
+---
+
+Originally posted as totw/130 on 2017-02-17
+
+By Titus Winters (titus@google.com)
+
+*The precision of naming takes away from the uniqueness of seeing* — Pierre
+Bonnard
+
+The earliest commit of the Google C++ Style Guide contains the guidance that
+many people are still using for namespace naming. Roughly, this can be
+summarized as "namespaces are derived from package paths," although there's an
+interesting less-followed option that an entire top-level package
+(search/...) could choose a single flat top-level namespace. Following on
+the heels of Java's package naming requirements, this makes a lot of sense: we
+want to be able to uniquely identify symbols in C++ and we want there to be
+uniqueness and consistency in namespace choice.
+
+Except in actuality, we don't. We just didn't realize for almost a decade.
+
+## Name lookup
+
+Let's start with how name lookup works in C++ and how it's different from Java.
+
+``` c++
+namespace foo {
+namespace bar {
+void f() {
+  Baz b;
+}
+}
+}
+```
+
+In C++, lookup on an unqualified name (`Baz`) will search expanding scopes for
+a symbol of the same name: first in `f` (the function), then in `bar`, then in
+`foo`, then in the global namespace.
+
+
+In Java, there is no such thing as an unqualified symbol: either a symbol is
+a qualified name
+
+``` java
+public void f() {
+  com.google.foo.bar.Baz b = new com.google.foo.bar.Baz();
+}
+```
+
+or it is imported, either as a single package member or via wildcard:
+
+``` java
+import com.google.foo.bar.Baz;
+import com.google.foo.bar.*;
+```
+
+In no case is `Baz` looked for outside of the package that is explicitly
+provided: wildcards don't descend into child packages, nor is search extended
+into parent packages. As it turns out, the difference in how parent
+packages/namespaces is handled is fundamental to why structural namespace naming
+(making the namespace structure match the package hierarchy) is a mistake.
```

Comment:
As it turns out, **this** difference ...  is a mistake **(in C++)** (?)

Reply:
Hmm. It's an intricate sentence.

as it turns out
(the difference in how parent packages/namespaces are handled) is fundamental to
(why structural namespace naming (...)) is a mistake.

Suggestions welcome.

Reply:
I think you have to spell out what you are "differing" (Java and C++) and what is the "mistake" (C++).

I.e.:

"As it turns out, this difference in how parent packages/namespaces are handled within Java and C++ is fundamental to why structural namespace naming (making the namespace structure match the package hierarchy) is a mistake within C++."

Reply:
Done.